### ISSUE TITLE
make it possible to unmount secrets

### ISSUE THREAD

**Issue by ibotty** [kubernetes/kubernetes#20151]
Many applications have a setup phase where they need credentials and don't afterwards. It would be great not to retain the secrets in case of an application vulnerability.

It might also be nice to do that automatically (e.g. after a readiness probe succeeds). I think it ought to be an attribute of the container's volume mount (and not the volume).

**Comment by thockin** [comment]
Could you instead put them into env vars (PRs are in or in flight, look for it in v1.2) and unset the env var when you are done?  The idea as described is not very exciting to me (readiness can change back and forth) and tricky to implement (crossing mount namespaces).

**Comment by ibotty** [comment]
I understand the code complexity argument.

I don't think, environment variables are the answer though. Coming from traditional Unix, I mistrust env vars for storing secrets, because they get exposed in proc. They also get shown when exporting the pod/rc declaration in kubernetes.

**Comment by thockin** [comment]
On Thu, Jan 28, 2016 at 9:58 PM, Tobias Florek <notifications@github.com> wrote:
>
> I understand the code complexity argument.
>
> I don't think, environment variables are the answer though. Coming from traditional Unix, I mistrust env vars for storing secrets, because they get exposed in proc. They also get shown when exporting the pod/rc declaration in kubernetes.

Using a secret as an env var is new in kube 1.2.  It doesn't show up
in your pod resource because it is late-bound by the kubelet.  Anyone
who can read your /proc/pid/environ can also read your files or ptrace
your process.

**Comment by hach-que** [comment]
This is important for security purposes to limit the attack surface of gaining credentials from the filesystem. Is there going to be any progress on this?

Our current mitigation strategy here is to encrypt the secrets before storing them in Kubernetes, and arm the Docker image with that encryption key. Then on startup, read the encrypted secrets from Kubernetes, decrypt them with the local encryption key, and then remove that key from the pod filesystem.

**Comment by thockin** [comment]
Can you say why env vars are not sufficient?

**Comment by hach-que** [comment]
My understanding is that environment variables are visible to other processes via `/proc`, so we definitely don't want to be using them if they're visible to other processes.

**Comment by thockin** [comment]
Only to other processes in the same PID space or the root PID space who run
as root or the same UID.  You can verify this by trying to open someone
else's /proc/<pid>/environment file.  It should fail unless you have
permissions.

It would also be visible through `docker inspect`, which also requires
permissions.

**Comment by hach-que** [comment]
Would the environment variables persist in those places (e.g. `docker inspect`), even if the process clears the environment variables internally (i.e. with `Environment.SetEnvironmentVariable` in .NET Core apps)?

**Comment by thockin** [comment]
Yes for `docker inspect` - it is part of the container's specification,
just as the secret exists in the kube API.  I don't actually know if /proc
reflects that sort of change, I have never tried (I am on mobile now), but
it should be easy to test.

But env bars are about as private as you are going to get without taking
full control and fetching secrets yourself into your own process memory.

**Comment by hach-que** [comment]
How is reading a file that exists on disk while the application is not serving traffic to memory, not more secure than an environment variable which persists around in `docker inspect`?

**Comment by thockin** [comment]
Its perms on disk are approximately the same as your own /proc files, so
not really more secure, but you seemed concerned about files.

There isn't a way to unmount Secrets (in fact, it will be refreshed
periodically).

I guess I don't know what your hypothetical attack vector is, so I can't
answer the question properly.

**Comment by hach-que** [comment]
Sure, it's permissions are approximately the same as the `/proc` files. But it exists only during startup, whereas `docker inspect` exists for the lifetime of the container.

If Kubernetes supported unmounting of secrets, then after startup the secret would exist only in the process's memory and nowhere else.

The key point here is to limit the amount of time the credential is readily available to someone who wants to read it. If it's in process memory, then an attacker would need to dump the memory of the process to disk, find the secret in whatever form it exists, and then send it over the network. That's a lot more difficult than `curl`'ing some string or file to a remote server.

**Comment by thockin** [comment]
Fair enough, though the short answer is that we don't support that for now,
and as far as I know nobody has proposed it.

**Comment by hach-que** [comment]
This is an important feature request for improved security in Kubernetes containers.

**Comment by tallclair** [comment]
https://diogomonica.com/2017/03/27/why-you-shouldnt-use-env-variables-for-secret-data/
Leaking env vars through debug endpoints is a common attack vector

There are a few tools you could use to accomplish this today:
- Perform the operations that need the secrets in an init container (if possible)
- Startup as root, and drop to an unprivileged user after the initialization (a good practice anyway, when possible)
- Use AppArmor change hat (requires deploying custom AppArmor profiles)

**Comment by MikeCamel** [comment]
I would vote against the AppArmor approach, though I'm also a big fan of _not_ using Env variables.  First, not all distributions will be running AppArmor, and second (most important), this pushes the problem out to another component, rather than fixing it within K8s itself.

**Comment by thockin** [comment]
If we wanted to implement this, an option might be for kubelet to open a
named pipe and write the secret value to it once.  The.pod could read it
once.

The problem is that a pod process might crash and restart - are we supposed
to re-offer the secret?

That's why this problem needs design :)

**Comment by XciD** [comment]
Hello,

I think is an important feature to consider in Kubernetes. 

How can we help ?

**Comment by thockin** [comment]
This problem needs design work - if you think it's important, you can start
thinking through the lifecycle of secrets in the face of things like pod
crash-restart cycles, in terms of API (how do I actually make the secret go
away), and with regards to features like subPath which could pin a secret's
inode.

**Comment by slayer** [comment]
I agree, that feature can seriously increase security of containerized application to store secrets only in memory of application, not in the environments or files since they do need after container start

**Comment by yuli-lendbuzz** [comment]
Hi guys.

This is an important feature that will greatly improve the container security.

Best, Yuli

**Comment by thockin** [comment]
https://github.com/kubernetes/kubernetes/issues/20151#issuecomment-562369295

**Comment by icu5545** [comment]
I, too, agree that this would undoubtedly be a useful feature.

@thockin I realize that the design needs to be thought out. However, from the messages above, it would seem that you're relying on the requesters to provide that design. That may be way off base, but that's how I've read this thread. I think, generally speaking, if the requesters had a design in mind, they may have made that known by now and/or forked the code and provided an implementation.

The fact is that the vast majority of those who ask for features are not experts in the implementation or details. Rather, they are end-users who are utilizing the product.

What I'm trying to get at here is that it would seem that those maintaining the project would be the ones defining the design for feature requests deemed worthy of inclusion. It may require redesign or reworking of other features that those project maintainers are more intimately familiar with. Sure, take any advice from the community that can be given, but maybe don't rely on the small chunk of people who happen to find this issue to provide all the details.

Again, I may be way off and you are, indeed, trying to come up with a design on your own. If you are, perhaps a quick acknowledgment of the fact or summary of your current thoughts would be helpful here. If not, is there anything to be done outside of waiting on community-driven designs?

**Comment by yuli-lendbuzz** [comment]
Hello @thockin and @icu5545 

I am not a kubernetes developer. But, I think I do have a simple solution.

Instead of secrets mounted as system volumes today, you can copy secretes values as files to the starting container.

For example, today, I can easily run ```kubectl cp``` to copy a secret file to the running container.

Do the same in automatic way on container start.

The script running in container will remove this secret file from file system when not needed.

Best
Yuli

**Comment by yuli-lendbuzz** [comment]
I was checking Hashicorp Vault to inject secrets for my containers.

I **can** delete secret files injected by Hashicorp Vault.

BUT, Hashicorp Vault implementation is not perfect.

On scale up / down, sometimes the secret files are restored.

I reported a bug: https://github.com/hashicorp/vault/issues/13676

**Comment by thockin** [comment]
Let's converge this into #86186
